cosim/black-parrot-example/vivado/tcl/constraints.tcl

```diff
@@ -0,0 +1,98 @@
+#       -source [get_pins rtc_mux2/I0] [get_pins rtc_mux2/O]
+#   create_generated_clock -name rtc_mux2_src2 -divide_by 1 -add -master_clock clk4 \
+#       -source [get_pins rtc_mux2/I1] [get_pins rtc_mux2/O]
+#   set_clock_groups -physically_exclusive -group rtc_mux2_src1 -group rtc_mux2_src2
+#
+#   create_generated_clock -name rtc_mux3_src1 -divide_by 1 -add -master_clock rtc_mux1_src1 \
+#       -source [get_pins rtc_mux3/I0] [get_pins rtc_mux3/O]
+#   create_generated_clock -name rtc_mux3_src2 -divide_by 1 -add -master_clock rtc_mux1_src2 \
+#       -source [get_pins rtc_mux3/I0] [get_pins rtc_mux3/O]
+#   create_generated_clock -name rtc_mux3_src3 -divide_by 1 -add -master_clock rtc_mux2_src1 \
+#       -source [get_pins rtc_mux3/I1] [get_pins rtc_mux3/O]
+#   create_generated_clock -name rtc_mux3_src4 -divide_by 1 -add -master_clock rtc_mux2_src2 \
+#       -source [get_pins rtc_mux3/I1] [get_pins rtc_mux3/O]
+#   set_clock_groups -physically_exclusive -group rtc_mux3_src1 -group rtc_mux3_src2 -group rtc_mux3_src3 -group rtc_mux3_src4
+#
+# However, since
+#   1. the w_reset_i pin of mtime_gray is wired to reset_i
+#   2. the select for the 4-1 mux is initialized to 2'b00 when reset_i is high
+#   3. clk_i is always faster than rt_clk_i
+# , we should only consider clk_i for mtime clock. In order to do this, instead of creating all the generated clocks like
+# the one in the above example, we only create necessary clocks for clk_i:
+# clk_i running through rtc_mux1
+create_generated_clock -name rtc_mux1_clk1 -divide_by 1 -source [get_pins $rtc_mux1/I0] -add -master_clock [get_clocks clk_fpga_0] [get_pins $rtc_mux1/O]
+# clk_i running through rtc_mux3
+create_generated_clock -name rtc_mux3_clk1 -divide_by 1 -source [get_pins $rtc_mux3/I0] -add -master_clock [get_clocks rtc_mux1_clk1] [get_pins $rtc_mux3/O]
+
+# max delay for the bufgmux_ctrl primitives
+set_max_delay -from [get_pins $clint_inst/mtimesel_reg/data_r_reg*/C] -to [get_pins $rtc_mux1/S*] $bp_min_period
+set_max_delay -from [get_pins $clint_inst/mtimesel_reg/data_r_reg*/C] -to [get_pins $rtc_mux2/S*] $bp_min_period
+set_max_delay -from [get_pins $clint_inst/mtimesel_reg/data_r_reg*/C] -to [get_pins $rtc_mux3/S*] $bp_min_period
+
+set_false_path -hold -from [get_pins $clint_inst/mtimesel_reg/data_r_reg*/C] -to [get_pins $rtc_mux1/S*]
+set_false_path -hold -from [get_pins $clint_inst/mtimesel_reg/data_r_reg*/C] -to [get_pins $rtc_mux2/S*]
+set_false_path -hold -from [get_pins $clint_inst/mtimesel_reg/data_r_reg*/C] -to [get_pins $rtc_mux3/S*]
+
+################# bsg_launch_sync_sync #################
+foreach blss_inst [get_cells -hier -filter {(ORIG_REF_NAME == bsg_launch_sync_sync || REF_NAME == bsg_launch_sync_sync)}] {
+  puts "blss_inst: $blss_inst"
+  #foreach launch_reg [get_cells -regexp {$blss_inst/.*/bsg_SYNC_LNCH_r_reg\\[.*]}]
+  foreach launch_reg [get_cells -regexp [format {%s/.*/bsg_SYNC_LNCH_r_reg\\[.*]} $blss_inst]] {
```

Comment:
Usually better to avoid hold paths by setting false path to the first sync e.g.

```
proc bsg_sync_constraints {} {
  foreach_in_collection s1 [get_pins -quiet -of_objects [get_cells -quiet -hier *hard_sync_int1_BSG_SYNC*] -filter "name=~D"] {
    set_false_path -to $s1 -setup
    set_false_path -to $s1 -hold
  }
}
```